File: BudgetTracker.API/Models/ExpenseModel.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace BudgetTracker.API.Models
{
    public class ExpenseModel
    {
        [Key]
        public int ExpenseID { get; set; }

        [Required(ErrorMessage = "O campo Título é obrigatório.")]
        public string? Title { get; set; }
        public string? Description { get; set; }

        [Required(ErrorMessage = "O campo Valor é obrigatório.")]
        [Range(0, double.MaxValue, ErrorMessage = "O valor deve ser maior que zero.")]
        public double Amount { get; set; }

        [Required(ErrorMessage = "O campo Data é obrigatório.")]
        public DateTime Date { get; set; }

        [ForeignKey("CategoryID")]
        public int CategoryID { get; set; } // Chave estrangeira

        public CategoryModel? Category { get; set; } // Propriedade de navegação
    }
}

File: budget_tracker/lib/models/expense_model.dart
import 'package:budget_tracker/models/category_model.dart';

class ExpenseModel {
  int? expenseID;
  String? title;
  String? description;
  double? amount;
  DateTime date;
  CategoryModel? category;

  ExpenseModel({
    this.expenseID,
    this.title,
    this.description,
    required this.amount,
    required this.date,
    required this.category,
  });

  Map<String, dynamic> toJson() {
    return {
      'expenseID': expenseID,
      'title': title,
      'description': description,
      'amount': amount!.toDouble(),
      'date': date.toIso8601String(),
      'category': category?.toJson(), // Serializar o objeto da categoria
    };
  }

  factory ExpenseModel.fromJson(Map<String, dynamic> json) {
    return ExpenseModel(
      expenseID: json['expenseID'],
      title: json['title'],
      description: json['description'],
      amount: json['amount'].toDouble(),
      date: DateTime.parse(json['date']),
      category: CategoryModel.fromJson(json['category']
          as Map<String, dynamic>), // Desserializar o objeto da categoria
    );
  }
}
